### ISSUE TITLE
Attachment Open Issue in NO7.0 and NO7.1

### ISSUE THREAD

**Issue by mail2mhossain** [NakedObjectsGroup/NakedObjectsFramework#36]
In both NO7.0 and NO7.1, attachment cannot be opened. Please find the sample project. Attached “NO7Db” and change your connection string. After rebuild, Run the project and enter admin@gmail.com in email field and “123456” in password field. Click on Area -> Division -> Show All Divisions. Division list will be displayed. Click on “Dhaka Division” and you will find an attachment. Click on the attachment. Attachment is not downloading/opening, error occurred.
[NO7TestSolution.zip](https://github.com/NakedObjectsGroup/NakedObjectsFramework/files/429802/NO7TestSolution.zip)

**Comment by richardpawson** [comment]
Thanks for posting the solution.  I'm afraid I won't be able to look at it until later in the week, as my priorities are on NOF 8 (and NOF 9 now under development!).  However, just a comment: given that NOF7 has been released for over a year and I know people are using it with attachments  -  is there something different about what you are doing e.g. the mime type of the attachement and/or the way that the content is being stored?

**Comment by mail2mhossain** [comment]
As per docoment, our Attachment class is 
#region Primitive Properties
        [Key, NakedObjectsIgnore]
        public virtual long FileAttachmentId { get; set; }
        [NakedObjectsIgnore]
        public virtual byte[] FileContent { get; set; }
        [Required, StringLength(200), NakedObjectsIgnore]
        [MemberOrder(10)]
        public virtual string FileName { get; set; }
        [NakedObjectsIgnore]
        public virtual string FileMime { get; set; }

        #endregion
        [Title]
        public virtual FileAttachment FileAttachment
        {
            get
            {
                if (FileContent == null) return null;
                return new FileAttachment(FileContent, FileName, FileMime) { DispositionType = "inline" };
            }
        }

        #region  Navigation Properties
        [Required, NakedObjectsIgnore]
        [MemberOrder(200)]
        public virtual Area Area { get; set; }
        #endregion

AND Get property is in Area class
#region Get Properties
        [NotMapped]
        [Eagerly(EagerlyAttribute.Do.Rendering)]
        [TableView(false, "FileAttachment")]
        public IList<Attachment> Attachments
        {
            get
            {
                IList<Attachment> attachments =
                    Container.Instances<Attachment>().Where(w => w.Area.AreaId == this.AreaId).ToList();
                return attachments;
            }
        }
        #endregion

**Comment by richardpawson** [comment]
I have looked into this briefly.  The stack trace on the error shows that it is something to do with authorization (the whole stack trace is Microsoft code - not NOF code):

```
System.Web.Mvc.AuthorizeAttribute.AuthorizeCore(HttpContextBase httpContext) +31
   System.Web.Mvc.AuthorizeAttribute.OnAuthorization(AuthorizationContext filterContext) +160
   System.Web.Mvc.ControllerActionInvoker.InvokeAuthorizationFilters(ControllerContext controllerContext, IList`1 filters, ActionDescriptor actionDescriptor) +97
etc.
```
The NOF controller method that should be being called when you click on an attachment is: GenericController #GetFile  -  however this isn't being called because the authorization code is failing first.

I added [OverrideAuthorization] onto just this GetFile method, and your example then worked fine.  I realise this isn't a proper solution, but it clarifies the issue.

I don't have time to investigate this further at present, but hopefully this gives you a start point for further investigation -  something in the bowels of the authorization code, it looks like. You will probably need to get the source for MVC and/or the authorization logic you are using and trace through.
